File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int prefixSum=0;
        Map<Integer, Integer> map=new HashMap<>(); // <sum, freq>

        // default
        map.put(0,1);

        // result
        int totalSub=0;

        for (int elem : nums) {
            prefixSum += elem;
            int complement = prefixSum - k;
            if (map.containsKey(complement)) {
                totalSub += map.get(complement);
            }
            map.put( prefixSum, map.getOrDefault(prefixSum, 0) + 1 );
        }

        return totalSub;
    }
}

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
# LC: 560
class Solution:
    def subarraySum(self, nums: List[int], K: int) -> int:
        prefix_sum = 0
        mapp = defaultdict(int) # <sum, freq>
        
        mapp[0] = 1
        found = 0

        for num in nums:
            prefix_sum += num
            complement = prefix_sum - K
            found += mapp[complement]
            mapp[prefix_sum] +=1

        return found